# Notes

For each species, manual curation is necessary to produce a usable sample sheet (in most cases). These are processing notes on each species.

At the moment, there is no automated way to add publication data for each metadata file, and so this is manually tracked via Google Sheet.

### Workflow

To add a new species, you'll need to do the following:

First, run the setup blocks. Input and output of sra metadata should be done using the get_sra and write_metadata functions, which standardize formats. You'll need to change the paths to point to the correct location on your computer of these files. **to change**

The get_sra function takes a BioProject ID (or species, but better to use BioProjects in most cases), and reads in the default columns from the sample metadata file (BioProjet, BioSample, Organism, sex). You'll likely want to read in additional columns, but which changes from dataset to dataset. Generally I try to capture location / population information, but for the moment have not been worrying about much else, except fields I'll need to filter on later (to remove domesticated samples, pooled samples, etc). 

It helps to look at the metadata file in Excel or a text editor to see what may be useful to read in, and also check the original paper to see what metadata they provide. Once you know what you want, specify the additional columns as a character vector to the opt_cols argument.

Now, you'll have the metadata from SRA. In some cases, you are done. In other cases you'll need to parse supplemental data from papers. The readxl (Excel spreadsheets), tabulizer (PDF tables), and rvest (html) packages are helpful here; the goal is that you should not need to manually touch anything. I try to use PMC links rather than journal links as they are probably more stable. 

Then you'll need to merge the sra and paper metadata, which can be tricky. If you really can't figure it out, and the SRA metadata is not sufficient on its own (the paper indicates multiple populations were sampled but you cannot link those populations to BioSamples), indicate metadata is inadequate and move on. 

Finally, you'll need to do some last cleanup and export. The Organism field is key here, as the export code splits on this field to produce one file per species. Check to make sure this field is correct (some datasets have a separate "Species" field or something that is more accuracte than Organism). Anything only identified to the genus level (e.g., Genus cf. species; Genus spp; Genus) should be renamed to Genus_spp, and all bionomials should be renamed to replace spaces with underscores (since these will become filenames). Then, run the export code, write_metadata. This splits the input dataframe by Organism, and adds a reference genome column based on the reference species indicated in the sp argument and the genome_key file.

All done! Don't forget to git commit and git push, with a helpful commit message, such as "Added metadata for species xxxx".

### Setup

First we load libraries

```{r}
library(rvest)
library(stringr)
library(tabulizer)
library(readxl)
library(tidyverse)

```

Then we set some default processing. For all datasets, we need the BioProject (to link to publication), the BioSample (to link to VCF), and Organism. 

Paths for input and output:

```{r}
clean_metadata_path <- "~/Projects/popgen/compPopGen_ms/SRA/cleaned-metadata/"
raw_metadata_path <- "~/Projects/popgen/compPopGen_ms/SRA/SRA-sample-metadata/"
```

A genome key file that links reference species to GenBank accession:

```{r}
genome_key_file <- "~/Projects/popgen/compPopGen_ms/SRA/Organism_Metadata.tsv"
genome_key <- read_tsv(file=genome_key_file)
```

Functions for reading and writing metadata:

```{r}

get_sra <- function(sp, path=raw_metadata_path, opt_cols) {
  default_cols = c("BioProject", "BioSample", "Organism", "sex")
  if (missing(opt_cols)) {
    cols = enquo(default_cols)
  } else {
    default_cols = c(default_cols, opt_cols)
    cols = enquo(default_cols)
  }
  read_tsv(str_c(raw_metadata_path, "SRA_Metadata_", sp, ".tsv")) %>% 
    select(!!cols)
}

write_metadata <- function(df, refSp) {
  refGenome <- genome_key %>% filter(Organism == refSp) %>% pull(AssemblyAccession)
  df %>% mutate(refGenome = refGenome) %>%
  split(., .$Organism) %>%
  imap(~ write_csv(as.data.frame(.x), file = str_c(clean_metadata_path, .y, '_metadata.csv')))
}
  
```

## METADATA CODE BELOW ##

### Amphilophus citrinellus

Sample metadata downloaded from: https://datadryad.org/stash/dataset/doi:10.5061/dryad.bcc2fqz91

Keyed on sample it / submitter id.


```{r}
sp<-"Amphilophus_citrinellus"
bp<-"PRJEB38173"

#metadata from paper
dryad_link<-"https://datadryad.org/stash/downloads/file_stream/561311"
paper<-read_csv(dryad_link) %>% select(Sample_name, Species, Lake, Sampling_Year)

sra<-get_sra(sp, opt_cols = c("Submitter_Id")) %>%
  mutate(Submitter_Id = str_remove(Submitter_Id, "_WGS")) %>% 
  inner_join(paper,by=c("Submitter_Id" = "Sample_name")) %>%
  mutate(Organism = case_when(
    Species == "Amphilophus cf. citrinellus" ~ "Amphilophus spp",
    Species == "Amphilophus spp. (hybrid)" ~ "Amphilophus spp",
    TRUE ~ Species
   )) %>%
  select(-Species) %>%
  mutate(Organism = str_replace_all(Organism, " ", "_"))

write_metadata(sra, sp)
```

### Anas platyrhynchos

Sample metadata embedded in PDF in supplemental material: https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6050300/bin/41467_2018_4868_MOESM1_ESM.pdf

Location of mallard sampling taken from paper text.

```{r}
sp="Anas_platyrhynchos"
sra<-get_sra(sp, opt_cols = c("Sample_Type", "sex")) %>% 
  mutate(Sample_Type = str_remove(Sample_Type, "_F"))

supmat<-"https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6050300/bin/41467_2018_4868_MOESM1_ESM.pdf"
paper<-do.call(rbind,extract_tables(supmat, pages=c(38:40)))
colnames(paper) <- c("sample", "breed", "mapping_rate", "insert_size", "depth")

meta<-paper %>% as_tibble %>% select(sample, breed) %>% 
  filter(breed != "Breed", breed != "") %>% 
  left_join(sra, by=c("sample" = "Sample_Type")) %>%
  filter(breed == "Mallard") %>%
  mutate(Continent = "Asia", Country = "China", Locality = "Aoji Duck Farm, Zhejiang Province")

write_metadata(meta, sp)

```

### Anguilla anguilla

Metadata provided to SRA is great and totally complete, no need to dig into paper.
Generally this is very low coverage data, with only a fraction of individuals above 5x, so may not be great.
Closely related species Anguilla rostrata also available

```{r}
sp <- "Anguilla_anguilla"
bp <- "PRJNA668259"

sra<-get_sra(bp, opt_cols = c("Isolate", "Ecotype", "Collection_date", "lat_lon", 
                              "geo_loc_name_country", "geo_loc_name_country_continent", 
                              "geo_loc_name")) %>%
  rename(Country=geo_loc_name_country, 
         Continent=geo_loc_name_country_continent,
         Locality=geo_loc_name)

write_metadata(sra, sp)
```

### Anolis carolinensis

Single individuals of A. porcatus and A. allisoni in same BioProject.

Metadata is okay, but need to get more info about location from the paper: https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6681179/

Locked in a Zip archive so need to process

```{r}
sp <- "Anolis_carolinensis"
bp <- "PRJNA533001"
  
sra<-get_sra(bp, opt_cols = c("Sample Name", "Sample_Type")) %>% 
  rename(Submitter_Id = `Sample Name`, Isolate = Sample_Type)

supmat<-"https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6681179/bin/evz110_supplementary_data.zip"
temp <- tempfile()
download.file(supmat,temp)
unzip(temp, files=c("Table_S1.xlsx"), junkpaths=TRUE, exdir=raw_metadata_path)
paper<-read_excel(str_c(raw_metadata_path, "Table_S1.xlsx")) %>% select(Sample:Longitude)
unlink(temp)

meta<-inner_join(paper, sra, by=c("Sample" = "Submitter_Id"))

write_metadata(meta, sp)

```

### Astatotilapia calliptera

This dataset is complex, spanning many species in the cichild flock. Best to process by BioProject, not species. The metadata is also terrible and poorly organized / hard to parse.

Will need to think a bit about whether this is usable, but the metadata is here if needed.

```{r}
sp="Astatotilapia_calliptera"
bp="PRJEB1254"

split_biosample <- function(text) {
   seq2 <- Vectorize(seq.default, vectorize.args = c("from", "to"))
   split_input <- as_tibble(str_split(text, "-", simplify = TRUE)) %>% 
     mutate(V2 = ifelse(V2 == "", V1, V2)) %>%
     mutate(output = sapply(seq2(from=str_remove(V1, "SAMEA"), to=str_remove(V2, "SAMEA"), by=1), paste, sep="", collapse=",")) %>%
     pull(output)
 }


sra<-get_sra(bp, opt_cols = c("Alias")) %>% 
  rename(Submitter_Id = Alias)

supmat1<-"https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4700518/bin/NIHMS66047-supplement-01.pdf"
table16<-do.call(rbind,extract_tables(supmat1, pages=c(47:47)))
colnames(table16) <- c("SampleName", "BioSample")

paper1<-as_tibble(table16) %>% separate_rows(BioSample, sep=",") %>% filter(SampleName != "Samples") %>%
  filter(BioSample != "") %>%
  separate_rows(BioSample, sep=",") %>%
  mutate(BioSample = str_remove(BioSample, "\\r")) %>%
  mutate(BioSample = trimws(BioSample)) %>%
  mutate(BioSample = split_biosample(BioSample)) %>%
  separate_rows(BioSample) %>% select(BioSample, SampleName) %>%
  mutate(BioSample = str_c("SAMEA", BioSample))

supmat2<-"https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6443041/bin/NIHMS82209-supplement-Supplementary_information.pdf"
table4 <- do.call(rbind,extract_tables(supmat2, pages=c(50:50)))
colnames(table4) <- c("SampleName", "BioSample")

paper2<-as_tibble(table4) %>% filter(SampleName != "Samples") %>%
  mutate(SampleNameClean = c(rep("Lake Malawi populations",11), rep("Lake Malawi trios",3), 
                        rep("Malawi catchment", 2), rep("Indian ocean catchments", 2),
                        rep("Outgroup", 6))) %>%
  filter(BioSample != "") %>%
  separate_rows(BioSample, sep=",") %>%
  mutate(BioSample = str_remove(BioSample, "A. stuartgranti")) %>%
  mutate(BioSample = str_remove(BioSample, "L. lethrinus")) %>%
  mutate(BioSample = str_remove(BioSample, "A. calliptera Salima")) %>%
  mutate(BioSample = trimws(BioSample)) %>%
  filter(str_detect(BioSample, "SAMEA")) %>%
  mutate(BioSample = str_replace(BioSample, "SAMEA1904329-", "SAMEA1904329-SAMEA1904331")) %>%
  mutate(BioSample = str_replace(BioSample, "SAMEA1904350-", "SAMEA1904350-SAMEA1904353")) %>%
  mutate(BioSample = trimws(BioSample)) %>%
  mutate(BioSample = split_biosample(BioSample)) %>%
  separate_rows(BioSample) %>% select(BioSample, SampleNameClean) %>%
  mutate(BioSample = str_c("SAMEA", BioSample))

meta<-full_join(paper1, paper2, by=c("BioSample" = "BioSample")) %>% distinct() %>% 
  full_join(sra, by=c("BioSample" = "BioSample")) %>% distinct() %>%
  rename(SampleName1 = SampleName, SampleName2 = SampleNameClean) %>%
  filter(!is.na(BioProject))

meta %>% filter(Organism == "Astatotilapia_calliptera" | Organism == "Astatotilapia") %>%
  mutate(Organism = ifelse(Organism == "Astatotilapia", "Astatotilapia_spp", Organism)) %>%
  write_metadata(sp)
```


### Astyanax mexicanus

Beautiful metadata, everything in SRA submission

```{r}
sp="Astyanax_mexicanus"
bp<-"PRJNA260715"
  
sra<-get_sra(bp, opt_cols = c("Sample Name")) %>% 
  rename(Submitter_Id = `Sample Name`) %>%
  separate(Submitter_Id, into=c("Population"), remove=FALSE, extra="drop")

write_metadata(sra, sp)

```

### Athene cunicularia

SRA submission mostly looks good, although there does not appear to be a way to link the SRA populations (A, B, C) to the real world places they are from.

```{r}

sp="Athene_cunicularia"
bp="PRJNA431202"
  
sra<-get_sra(bp, opt_cols = c("Sample Name", "Ecotype")) %>% 
  rename(Submitter_Id = `Sample Name`) %>%
  separate(Ecotype, into=c("Site", "Site_Type", "Year"), remove=FALSE, extra="drop")

write_metadata(sra, sp)
```

### Chaenogobius annularis

Adding information about population defined in the paper (Pacific Ocean vs Sea of Japan). 

Japan:Misaki -> PO
Japan:Misaki -> PO
Japan:Onagawa -> PO
Japan:Onagawa -> PO
Japan:Aomori -> SJ
Japan:Aomori -> SJ
Japan:Aomori -> SJ
Japan:Murakami -> SJ
Japan:Murakami -> SJ
Japan:Murakami -> SJ
Japan:Misaki -> PO
Japan:Onagawa -> PO

(from Table S4 from the paper)

```{r}

sp="Chaenogobius_annularis"
bp="PRJDB7125"
  
sra<-get_sra(bp, opt_cols = c("Isolate", "Collection_date", "geo_loc_name_country", "geo_loc_name_country_continent", "geo_loc_name", "Sample_Name")) %>% 
  rename(Submitter_Id = Sample_Name, Country = geo_loc_name_country, Continent = geo_loc_name_country_continent, Locality = geo_loc_name) %>%
  mutate(Population = case_when(
    Locality == "Japan:Misaki" | Locality == "Japan:Onagawa" ~ "Pacific Ocean",
    Locality == "Japan:Aomori" | Locality == "Japan:Murakami" ~ "Sea of Japan"    
  ))

write_metadata(sra, sp)

```


### Chloebia gouldiae

Very nice SRA metadata

```{r}

sp="Chloebia_gouldiae"
bp="PRJNA515277"
  
sra<-get_sra(bp, opt_cols = c("Ecotype", "Collection_date", "geo_loc_name_country", "geo_loc_name_country_continent", "geo_loc_name", "lat_lon", "Strain", "Cultivar")) %>% 
  rename(Submitter_Id = Cultivar, Country = geo_loc_name_country, Continent = geo_loc_name_country_continent, Locality = geo_loc_name) %>%
  select(-sex) %>%
  rename(sex = Strain)

write_metadata(sra, sp)

```

### Clupea spp.

Complicated by the fact that different metadata fields are used for different BioProjects; will probably need to parse by BioProject and then combined to avoid lots of complicated case_when statements.

Parse by BioProject, then output by species.

Additional metadata could be added froom "meta1" and "meta2" tables, but it is a confusing mess and I haven't done this yet.


```{r}

#PRJNA642736 filter to keep only Sample_Type == "individual"
sp="Clupea_harengus"

sra1<-get_sra("PRJNA642736", opt_cols = c("Library Name", "Ecotype", "Sample_Type")) %>% 
  rename(Submitter_Id = `Library Name`, Locality = Ecotype) %>%
  filter(Sample_Type == "individual") %>%
  select(-Sample_Type) %>% select(Organism, BioProject, BioSample, sex, Submitter_Id, Locality)

temp <- tempfile()
supmat<- "https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7738190/bin/elife-61076-supp2.xlsx"
download.file(supmat,temp)
meta1<-read_excel(temp, skip=1, col_types = "text", trim_ws = TRUE) 
unlink(temp)

#PRJNA338612 filter to keep only Sample_Type == "individual"

#grab pop info from paper

paper2<-"https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5410801/pdf/pnas.201617728.pdf"
table1 <- extract_tables(paper2, pages=c(3:3))[[2]]
colnames(table1) <- c("Locality", "N", "blank", "lat", "lon", "salinity", "date", "spawn_season")
meta2<-as_tibble(table1) %>% 
  filter(date != "") %>% 
  select(-blank, -N) %>% 
  separate(Locality, into=c("Locality", "Population"), remove=FALSE, sep="[†*]") %>% 
  mutate(Population = trimws(str_replace(Population, "[Ww]", ""))) %>%
  mutate(Population = str_replace(Population, "[Ä]", "A"))

sra2<-get_sra("PRJNA338612", opt_cols = c("Sample Name", "Sample_Type")) %>% 
  rename(Submitter_Id = `Sample Name`) %>%
  filter(Sample_Type == "Individual DNA") %>%
  select(-Sample_Type) %>% separate(Submitter_Id, into=c("Population", "Sample"), sep=-1, remove=FALSE) %>%
  left_join(meta2, by=c("Population" = "Population")) %>% select(Organism, BioProject, BioSample, sex, Submitter_Id, Locality)

sra3<-get_sra("PRJNA356817", opt_cols = c("Sample Name")) %>% 
  rename(Submitter_Id = `Sample Name`) %>%
  mutate(Locality = case_when(
    Submitter_Id == "AAPF" | Submitter_Id == "AAPM" ~ "Norway:Bergen",
    Submitter_Id == "ABPF" | Submitter_Id ==  "ABPM" ~ "Sweden:Hästskär",
    TRUE ~ NA_character_)) %>%
    filter(!is.na(Locality))  %>% select(Organism, BioProject, BioSample, sex, Submitter_Id, Locality)

#export
bind_rows(sra1, sra2, sra3) %>%
  write_metadata(sp)

```

### Coilia nasus

Sample metadata is in a word document, will need to find a way to parse that...



```{r}

sp="Coilia_nasus"
bp="PRJNA422339"

sra<-get_sra(bp)
```





